### PR TITLE
Allow exporting key material and init of cipherstate

### PR DESCRIPTION
**Pull request by aidantwoods-1p**
In order to allow implementations to store the nonce and key (e.g. for use with distributed systems), I've added an `UnsafeKey` method that exports the current key for storage. Combined with the `CipherSuite`, already exportable nonce, and this key, these can be used to reconstruct a `CipherState` from stored data using the new `UnsafeNewCipherState` method.

Noting that `UnsafeKey` isn't strictly necessary if it is already accessible from the `Cipher`, but I think this is preferable overall to needing to possibly create wrapped `Cipher` types which allow the key to be later exportable.

The `Unsafe` keyword is mean to indicate that these functions should be avoided at a glance, unless the caller accepts responsibility for manually managing exported state outside of application memory.

### CODE REVIEW

File: state.go [hunk 1]
```diff
@@ -32,6 +32,18 @@ const MaxNonce = uint64(math.MaxUint64) - 1
 var ErrMaxNonce = errors.New("noise: cipherstate has reached maximum n, a new handshake must be performed")
 var ErrCipherSuiteCopied = errors.New("noise: CipherSuite has been copied, state is invalid")
 
+// UnsafeNewCipherState allows a CipherState to be reconstructed from exported components.
```
Comment (titanous) [inline state.go:35]:
```suggestion
// UnsafeNewCipherState reconstructs a CipherState from exported components.
```

File: state.go [hunk 2]
```diff
@@ -32,6 +32,18 @@ const MaxNonce = uint64(math.MaxUint64) - 1
...
+// Note that it is important that, if resuming from exported state, then care is taken
```
Comment (titanous) [inline state.go:36]:
```suggestion
// It is important that, when resuming from an exported state, care is taken
```

File: state.go [hunk 3]
```diff
@@ -91,6 +103,13 @@ func (s *CipherState) SetNonce(n uint64) {
 	s.n = n
 }
 
+// UnsafeKey returns the current value of k. This can be used to export the current key
```
Comment (titanous) [inline state.go:106]:
```suggestion
// UnsafeKey returns the current value of k. This exports the current key
```

**Review by titanous — CHANGES_REQUESTED** [review]
This looks good, just minor nits in the docs.

**Comment by aidantwoods-1p** [comment]
Nits now resolved 🙂

**Review by titanous — APPROVED** [review]

**Comment by complexspaces** [comment]
Thank you for taking a look at this so quick for us 🎉

**Pull request merged**